### PR TITLE
[device-uptime] Uptime hourly calculation

### PR DESCRIPTION
**Pull request by ja-odur**
#### Summary of Changes (What does this PR do?)
- Enable hourly calculation of device and network uptime (can be customised too using the `MONITOR_FREQUENCY_MINUTES` env key)
- 
#### Is this change ready to hit production in its current state?
- [x] Yes
- [ ] No

#### Status of maturity (all need to be checked before merging):

- [x] I consider this code done
- [x] I've tested this locally
- [ ] This branch has been LGTM'ed by a reviewer

#### How should this be manually tested?
* Fetch this branch locally.
* Change directory to the `device-uptime` micro-service directory.
* Follow the steps in the `README.md` file to run the script.
* The script should run successfully.

#### What are the relevant tickets?
- [PLAT-626](https://airqoteam.atlassian.net/browse/PLAT-626)

### CODE REVIEW

**Review by NoahNsimbe — APPROVED** [review]
LGTM

File: src/device-uptime/utils/device_hourly_records.py [hunk 1]
```diff
@@ -60,56 +46,27 @@ def get_sensor_readings(self):
...
-        time = daily_data.index.tolist()
-
         return DeviceSensorReadings(
-            time=time,
-            sensor_one_pm2_5=sensor_one_pm2_5,
-            sensor_two_pm2_5=sensor_two_pm2_5,
-            battery_voltage=battery_voltage,
+            time=self.record.get("created_at"),
+            sensor_one_pm2_5=self.record.get("pm2_5"),
+            sensor_two_pm2_5=self.record.get("s2_om2_5"),
```
Comment (Baalmart) [inline src/device-uptime/utils/device_hourly_records.py:52]: please review this line, is this supposed to be s2_pm2_5? @ja-odur

File: src/device-uptime/utils/device_hourly_records.py [hunk 2]
```diff
@@ -60,56 +46,27 @@ def get_sensor_readings(self):
...
-        actual_records = self.get_record_count()
-        uptime = round((actual_records / expected_records) * percent_100, 2)
+        time = date_parser.isoparse(time)
+        now = datetime.utcnow()
+        now = now.replace(tzinfo=UTC)
+        minutes_diff = (now - time).total_seconds() / 60
 
-        if uptime > percent_100:
-            uptime = percent_100
+        if minutes_diff > configuration.MONITOR_FREQUENCY_MINUTES:
```
Comment (Baalmart) [inline src/device-uptime/utils/device_hourly_records.py:68]: Great addition here. :)

**Review by Baalmart — CHANGES_REQUESTED** [review]
All looks good, just one concern as shared in one of my comments @ja-odur

**Comment by Baalmart** [comment]
Awesome code organisation, I like the flat organisation of things with little to no need to search in mutliple subdirectories.

Also, my operation timeout on me while testing, please check out the screenshots. @ja-odur 

Screenshot 2021-06-24 at 18 52 53
Screenshot 2021-06-24 at 18 53 03

**Comment by Baalmart** [comment]
Since this job/micro interacts with some DBs like the one for device registry directly (Devices), we might also have to consider the ongoing Kafka works/connections in the future PRs. @NoahNsimbe

**Comment by NoahNsimbe** [comment]
> Since this job/micro interacts with some DBs like the one for device registry directly (Devices), we might also have to consider the ongoing Kafka works/connections in the future PRs. @NoahNsimbe

true, @Baalmart we shall discuss about it

File: src/device-uptime/utils/device_hourly_records.py (hunk 1 above)
Comment (ja-odur) [reply src/device-uptime/utils/device_hourly_records.py:52]: Nice catch @Baalmart

Comment (ja-odur) [reply src/device-uptime/utils/device_hourly_records.py]: @Baalmart This has been fixed

**Comment by ja-odur** [comment]
> This could be due to your internet connectivity, not the traceback is printed (it's not an actual exception), so for channel apis that timeout, the others are still processed. I have toned down the console color to avoid this confusion.

This could be due to your internet connectivity, not the traceback is printed (it's not an actual exception), so for channel apis that timeout, the others are still processed. I have toned down the console color to avoid this confusion.

**Review by Baalmart — APPROVED** [review]
LGTM

**Pull request merged**